src/sysupdate/sysupdate-transfer.c

```diff
@@ -0,0 +1,1282 @@
+                                return r;
+                        if (r == 0) {
+                                /* Child */
+
+                                const char *cmdline[] = {
+                                        "systemd-import",
+                                        "raw",
+                                        "--direct",          /* just copy/unpack the specified file, don't do anything else */
+                                        arg_sync ? "--sync=yes" : "--sync=no",
+                                        i->path,
+                                        t->temporary_path,
+                                        NULL
+                                };
+
+                                (void) unsetenv("NOTIFY_SOCKET");
+                                execv(import_binary_path(), (char *const*) cmdline);
+                                log_error_errno(errno, "Failed to execute %s tool: %m", import_binary_path());
+                                _exit(EXIT_FAILURE);
+                        }
+                        break;
+
+                case RESOURCE_PARTITION:
+
+                        /* regular file → partition */
+
+                        r = safe_fork("(sd-import-raw)", FORK_RESET_SIGNALS|FORK_DEATHSIG|FORK_LOG|FORK_WAIT, NULL);
+                        if (r < 0)
+                                return r;
+                        if (r == 0) {
+                                /* Child */
+
+                                const char *cmdline[] = {
+                                        "systemd-import",
+                                        "raw",
+                                        "--direct",          /* just copy/unpack the specified file, don't do anything else */
+                                        "--offset", offset,
+                                        "--size-max", max_size,
+                                        arg_sync ? "--sync=yes" : "--sync=no",
+                                        i->path,
+                                        t->target.path,
+                                        NULL
+                                };
+
+                                (void) unsetenv("NOTIFY_SOCKET");
+                                execv(import_binary_path(), (char *const*) cmdline);
+                                log_error_errno(errno, "Failed to execute %s tool: %m", import_binary_path());
+                                _exit(EXIT_FAILURE);
+                        }
+                        break;
+
+                default:
+                        assert_not_reached();
+                }
+
+                break;
+
+        case RESOURCE_DIRECTORY:
+        case RESOURCE_SUBVOLUME:
+                assert(IN_SET(t->target.type, RESOURCE_DIRECTORY, RESOURCE_SUBVOLUME));
+
+                /* directory/subvolume → directory/subvolume */
+
+                r = safe_fork("(sd-import-fs)", FORK_RESET_SIGNALS|FORK_DEATHSIG|FORK_LOG|FORK_WAIT, NULL);
```

Comment:
same pattern is repeated almost identically (different command line, fork name and binary_path, maybe add a static helper?

Reply:
indeed